### PR TITLE
Purge orphaned tip blocks on startup

### LINKED ISSUE #601: startup should not assume DBs on main chain
Since there is no reorg notification if not connected to dcrd during the reorg, starting up dcrdata cannot rely on an explicit signal from dcrd.

On startup, dcrdata should first check that the hash of it's best block matches what dcrd says is at that height.  If there is disagreement, dcrdata should self-trigger a reorg.

### PR DESCRIPTION
**Pull request by buck54321**
Uses the existing purge-n-blocks loop. `rpcutils.OrphanedTipLength` compares a database to dcrd from the tip height down until a common mainchain ancestor is found, and returns the number of sidechain blocks to be purged.

Resolves #601

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -255,8 +255,38 @@ func _main(ctx context.Context) error {
 		return
 	}
 
-	// Optionally purge best blocks according to config.
-	if cfg.PurgeNBestBlocks > 0 {
+	// Check for database tip blocks that have been orphaned. If any are found,
+	// purge blocks to get to a common ancestor. Only message when purging more
+	// than requested in the configuration settings.
+	blocksToPurge := cfg.PurgeNBestBlocks
```
Comment (chappjc) [inline main.go]: good solution.

File: rpcutils/rpcclient.go [hunk 2]
```diff
@@ -448,6 +449,53 @@ func CommonAncestor(client *rpcclient.Client, hashA, hashB chainhash.Hash) (*cha
 	return &hashA, chainA, chainB, nil
 }
 
+// OrphanedTipLength finds a common ancestor by iterating block heights
+// backwards until a common block hash is found. Unlike CommonAncestor, an
+// orphaned DB tip whose corresponding block is not known to dcrd will not cause
+// cause an error. The number of blocks that have been orphaned is returned.
+// Realistically, this should rarely be anything but 0 or 1, but no limits are
+// placed here on the number of blocks checked.
+func OrphanedTipLength(ctx context.Context, client *rpcclient.Client,
```
Comment (chappjc) [inline rpcutils/rpcclient.go]: Trying to think how this could bet tested via go test.  One idea is to replace `*rpcclient.Client` with `BlockHashGetter` so you can stub out the client.  I think it would be worth testing all the edge cases by manipulating what the `BlockHashGetter` and the `hashFunc` return.

File: rpcutils/rpcclient.go [hunk 3]
```diff
@@ -448,6 +449,53 @@ func CommonAncestor(client *rpcclient.Client, hashA, hashB chainhash.Hash) (*cha
...
+	commonHeight := tipHeight
+	var dbHash string
+	var err error
+	var dcrdHash *chainhash.Hash
+	for {
+		// Since there are no limits on the number of blocks scanned, allow
+		// cancellation for a clean exit.
+		select {
+		case <-ctx.Done():
+			return 0, nil
```
Comment (chappjc) [inline rpcutils/rpcclient.go]: If there is an interrupt, and this returns `0, nil` then I feel like there needs recognize this in main.  It could either be one of these:

```go
// Allow Ctrl-C to halt startup here.
	if shutdownRequested(ctx) {
return nil
```

or `OrphanedTipLength` could return -1 or a special non-nil error.

File: rpcutils/rpcclient.go (hunk 2 above)
Comment (buck54321) [reply rpcutils/rpcclient.go]: Thanks for the hint here. I already had the test written, but with a `rpcclient.Client` with my local dcrd credentials, so I didn't include it.

File: rpcutils/rpcclient_test.go [hunk 4]
```diff
@@ -114,3 +117,101 @@ func TestReverseStringSlice(t *testing.T) {
...
+	hashes = map[int64]string{
+		5: "something else",
+		4: "side block",
+		3: "blah",
+		2: "blue",
+		1: "flu",
+		0: "fly",
+	}
+
+	uncommon, err = OrphanedTipLength(ctx, client, 5, hashFunc)
```
Comment (chappjc) [inline rpcutils/rpcclient_test.go]: Just need to `_` out `uncommon`. (ineffassign)

**Comment by ZeroASIC** [comment]
I just tested this on a testnet database that was on the wrong chain and stuck and it seems to have updated everything correctly.  It was stuck on block 122934 so I checked blocks 122924-122934 after updating and everything looks the same as on testnet.dcrdata.org.
